doc/reference/trd-uart.md

```diff
@@ -0,0 +1,523 @@
+still partitioned into bytes, and the UART sends the least significant
+bits of each byte. Suppose a UART is configured to send 7-bit
+words. If a client sends 5 bytes, the UART will send 35 bits,
+transmitting the bottom 7 bits of each byte. The most significant bit
+of each byte is ignored.
+
+```rust
+pub enum StopBits {
+    One = 1,
+    Two = 2,
+}
+
+pub enum Parity {
+    None = 0,
+    Odd = 1,
+    Even = 2,
+}
+
+pub enum Width {
+    Six = 6,
+    Seven = 7,
+    Eight = 8,
+}
+
+pub struct Parameters {
+    pub baud_rate: u32, // baud rate in bit/s
+    pub width: Width,
+    pub parity: Parity,
+    pub stop_bits: StopBits,
+    pub hw_flow_control: bool,
+}
+
+pub trait Configuration {
+    fn get_baud_rate(&self) -> u32;
+    fn get_width(&self) -> Width;
+    fn get_parity(&self) -> Parity;
+    fn get_stop_bits(&self) -> StopBits;
+    fn get_flow_control(&self) -> bool;
+    fn get_configuration(&self) -> Configuration;
+}
+
+pub trait Configure {
+    fn set_baud_rate(&self, rate: u32) -> Result<u32, ErrorCode>;
+    fn set_width(&self, width: Width) -> Result<(), ErrorCode>;
+    fn set_parity(&self, parity: Parity) -> Result<(), ErrorCode>;
+    fn set_stop_bits(&self, stop: StopBits) -> Result<(), ErrorCode>;
+    fn set_flow_control(&self, on: bool) -> Result<(), ErrorCode>;
+    fn configure(&self, params: Parameters) -> Result<(), ErrorCode>;
+}
+```
+
+Methods in `Configure` can return the following error conditions:
+  - `OFF`: The underlying hardware is currently not available, perhaps
+    because it has not been initialized or in the case of a shared
+    hardware USART controller because it is set up for SPI.
+  - `INVAL`: Baud rate was set to 0.
+  - `ENOSUPPORT`: The underlying UART cannot satisfy this configuration.
+  - `FAIL`: Other failure condition.
+
+
+The UART may be unable to set the precise baud rate specified. For
+example, the UART may be driven off a fixed clock with integer
+prescalar. A call to `configure` MUST set the baud rate to the closest
+possible value to the `baud_rate` field of the `params` argument and a
+call to `set_baud_rate` MUST set the baud rate to the closest possible
+value to the `rate` argument. The `Ok` result of `set_baud_rate`
+includes the actual rate set, while an `Err(INVAL)` result means the
+requested rate is well outside the operating speed of the UART (e.g., 4MHz).
+
+
+3 `Transmit` and `TransmitClient`
+===============================
+
+The `Transmit` and `TransmitClient` traits allow a client to transmit
+bytes over the UART.
+
+```rust
+enum AbortResult {
+    Failure,
+    Success,
+}
+
+pub trait Transmit<'a> {
+    fn set_transmit_client(&self, client: &'a dyn TransmitClient);
+
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    fn transmit_word(&self, word: u32) -> Result<(), ErrorCode>;
```

Comment:
Should this 'no-buffer' path be `word` or `byte`? I can't think of a ton of scenarios where a caller wants to send exactly 4 bytes over UART?

Reply:
It's a word in case the UART is 9-bit data.